docs/rfcs/041-rel-sparse-keyspace.md

```diff
@@ -0,0 +1,176 @@
+
+We will use the recently introduced sparse keyspace to store actual data. Sparse keyspace was proposed in
+[038-aux-file-v2.md](038-aux-file-v2.md). The original reldir has one single value of `HashSet<(Oid, u8)>`
+for each of the databases (identified as `spcnode, dbnode`). We encode the `Oid` (`relnode, forknum`),
+into the key.
+
+```
+(REL_DIR_KEY_PREFIX, spcnode, dbnode, relnode, forknum, 1) -> deleted
+(REL_DIR_KEY_PREFIX, spcnode, dbnode, relnode, forknum, 1) -> exists
+```
+
+Assume all reldir data are stored in this new keyspace; the 3 reldir operations we mentioned before can be
+implemented as follows.
+
+1. Check if a relation exists: check if the key maps to "exists".
+2. List all relations: scan the sprase keyspace over the `rel_dir_key_prefix`. Extract relnode and forknum from the key.
+3. Create/drop a relation: write "exists" or "deleted" to the corresponding key of the relation.
+
+The mapping is implemented as `rel_tag_sparse_key` in the PoC patch.
+
+## Changes to Sparse Keyspace
+
+Previously, we only used sparse keyspaces for the aux files, which did not carry over when branching. The reldir
+information needs to be preserved from the parent branch to the child branch. Therefore, the read path needs
+to be updated accordingly to accommodate such "inherited sparse keys". This is done in
+[PR#10313](https://github.com/neondatabase/neon/pull/10313).
+
+## Coexistence of the Old and New Keyspaces
+
+Migrating to the new keyspace will be done gradually: when we flip a config item to enable the new reldir keyspace, the
+ingestion path will start to write to the new keyspace and the old reldir data will be kept in the old one. The read
+path needs to combine the data from both keyspaces.
+
+Theoretically, we could do a rewrite at the startup time that scans all relation directories and copies that data into the
+new keyspace. However, this could take a long time, especially if we have thousands of tenants doing the migration
+process simultaneously after the pageserver restarts. Therefore, we propose the coexistence strategy so that the
+migration can happen seamlessly and imposes no potential downtime for the user.
+
+With the coexistence assumption, the 3 reldir operations will be implemented as follows:
+
+1. Check if a relation exists
+  - Check the new keyspace if the key maps to any value. If it maps to "exists" or "deleted", directly
+    return it to the user.
+  - Otherwise, deserialize the old reldir key and get the result.
+2. List all relations: scan the sparse keyspace over the `rel_dir_key_prefix` and deserialize the old reldir key.
+   Combine them to obtain the final result.
+3. Create/drop a relation: write "exists" or "deleted" to the corresponding key of the relation into the new keyspace.
+  - Alternatively, when a user removes a relation when the new keyspace is enabled but it is created in the
+    old reldir key, we can remove it from the old reldir key and write the delete tombstone in the new keyspace.
```

Comment:
I don't think we have to do this. We can just trust the v2 mechanism:
If the relation is deleted in v2, but present in v1, the relation is indeed deleted.

Reply:
The problem is that we have to keep the delete tombstone in v2 forever unless we implement the "Full Migration" image layer generation strategy. i.e., the `create_image_layer` function cannot drop a v2 key if it is `deleted`. With this alternative trick we could remove these data early.

Considering relation creation/deletion doesn't sound like a frequent operation, maybe we don't need to spend too much engineering effort on garbage collection. We could probably just keep everything there.

Reply:
> the create_image_layer function cannot drop a v2 key if it is deleted

Otherwise, consider t1 gets created before the sparse keyspace is enabled and then deleted after the keyspace gets enabled. If create_image_layers simply drop the `deleted` key, t1 will become visible again.

Reply:
> Considering relation creation/deletion doesn't sound like a frequent operation, maybe we don't need to spend too much engineering effort on garbage collection. We could probably just keep everything there.

I don't think this assumption holds. We have users that use Neon e.g. for CI workflows, where each run will create and delete a bunch of relations. I'd be wary of replacing O(current relations) with O(historical relations) for listings.

Ideally, I think we should get rid of the tombstones and just delete the keys, but I understand that we'll need them before we're fully migrated. I think it's worth going through with a full migration to the sparse keyspace and get rid of the old relations and then revisit -- multiple sources of truth are a landmine waiting to blow up.

Reply:
In this case, I'd rather implement the write path as: if a key exists in the old keyspace, delete it from the old one; otherwise, put a delete tombstone in the new keyspace and clean it up during create_image_layers.

Reply:
I like that idea. It makes deletes O(reldir_v1), but that set will be monotonically decreasing, so it's not quadratic.

But in this case, why do we even need the tombstone? Is it only to speed up existence checks?

Reply:
Imagine we are now full opt-in to the new sparse keyspace. If a relation is created and then dropped, we need to record the state of "dropped" by writing a tombstone. On the storage level, we cannot "remove" a key in the history -- the way to remove it is by writing a tombstone and then collect it during image layer creation.

Reply:
Sure, but this is a KV tombstone from a delete operation, not a key with value "removed", right?

Reply:
Yes, but the system doesn't have native support for delete tombstone for now (the put_tombstone function is a no-op), and we need a key with value "removed" to represent a tombstone in the sparse keyspace.

Reply:
I see, thanks for clarifying.

Reply:
For the aux key space, we map empty value to tombstone. We can do the same for the relsize keyspace.